board/main.c

```diff
@@ -505,6 +490,98 @@ void __attribute__ ((noinline)) enable_fpu() {
   SCB->CPACR |= ((3UL << (10 * 2)) | (3UL << (11 * 2)));
 }
 
+uint64_t tcnt = 0;
+uint64_t marker = 0;
+
+// called once per second
+void TIM3_IRQHandler() {
+  #define CURRENT_THRESHOLD 0xF00
+  #define CLICKS 5 // 5 seconds to switch modes
+
+  if (TIM3->SR != 0) {
+    can_live = pending_can_live;
+
+    //puth(usart1_dma); puts(" "); puth(DMA2_Stream5->M0AR); puts(" "); puth(DMA2_Stream5->NDTR); puts("\n");
+
+    uint32_t current = adc_get(ADCCHAN_CURRENT);
+
+    switch (usb_power_mode) {
+      case USB_POWER_CLIENT:
+        if ((tcnt-marker) >= CLICKS) {
```

Comment:
minor thing and not related to this refactor: I would compute `(tcnt-marker) >= CLICKS` once, before L508 and collapse the if conditions
```
if (a) {
  if (b) {
```
into 
```
if (a && b) {
```